ethereumj-core/src/main/java/org/ethereum/vm/program/Program.java

```diff
@@ -1170,12 +1171,20 @@ public void callToPrecompiledAddress(MessageCall msg, PrecompiledContract contra
             track.rollback();
         } else {
 
-            this.refundGas(msg.getGas().longValue() - requiredGas, "call pre-compiled");
-            byte[] out = contract.execute(data);
+            Pair<Boolean, byte[]> out = contract.execute(data);
 
-            this.memorySave(msg.getOutDataOffs().intValue(), out);
-            this.stackPushOne();
-            track.commit();
+            if (out.getLeft()) { // success
+                this.refundGas(msg.getGas().longValue() - requiredGas, "call pre-compiled");
+                this.stackPushOne();
+                track.commit();
+            } else {
+                // spend all gas on failure, push zero and revert state changes
+                this.refundGas(0, "call pre-compiled");
+                this.stackPushZero();
+                track.rollback();
```

Comment:
M.b. refactor `PrecompiledContract` a bit and make it similar to regular call, I.e. leave only `execute()` there and throw OOGException if not enough gas for supplied data, and other custom exceptions if invalid data supplied (there you can supply a detailed description on why the call failed) ?

Reply:
Good idea, but let's make refactoring out of this PR. I spawned related issue https://github.com/ethereum/ethereumj/issues/926 for that